### ISSUE TITLE
Dot graph generated from workflow files

### ISSUE THREAD

**Issue by jackrabb1t** [Factual/drake#37]
If a dot graph could be generated from a workflow file, then, using graphwiz/gephi, we could get an intuitive image of the workflow

**Comment by mattgillooly** [comment]
Any progress on building this into Drake?  Here's a little Ruby script I tossed together to build a graph by parsing the output of `drake --print`, as suggested above.  https://gist.github.com/d007a2641b625cc92c6e

**Comment by dirtyvagabond** [comment]
@mattgillooly thanks for the reference script! 

What would you see as the main benefit of having Drake know how to do this, as opposed to the current decoupled approach? Are you worried about breaking changes; ease of use; something else?

**Comment by mattgillooly** [comment]
@dirtyvagabond - If you mean decoupling of the executables, that doesn't bother me so much.  Breaking changes to the `--print` interface are a bit of a concern.  I only found that option via this Issue thread, so I don't know how stable I can expect its output to be.  I'm sure you can speak to that.

If you mean complete decoupling of the projects, I'd like to push back a little.  I would like to see an "official" approach to this, and earlier comments gave me the impression that this was already underway 2 years ago.

Firstly, there's a discoverability concern.  I think other Drake users would find a Graphviz integration useful, but they are unlikely to find my Gist.  I wouldn't be surprised if someone else had followed the suggestion from 2 years ago, implemented a similar script, and simply neglected to share it here.

Additionally, while my script solves my own current problems, it is not intended as a general-purpose tool, so those who do find it may find it lacking:

* It only generates PNGs, not DOT.  I have no use for DOT immediately, but it's a perfectly reasonable thing to expect from a general purpose tool, e.g. for importing into Gephi as the original comment suggested.
* It ignores tags.  I don't use tags in Drake, so I'm not sure if this would matter to someone who does use them.
* It requires a newish Ruby + the ruby-graphviz gem installed.  Not a major hassle if you're familiar with Ruby, but it's one more thing that could prevent someone from using the tool.

These are things that make sense to fix if there is a coordinated effort around this functionality, but not so much if I'm the only one that will ever use it.

Again, I'm not sure which type of decoupling you meant, and you've got better visibility than I do into whether the Drake community would find this generally useful, so do with that braindump what you will.  :-)

**Comment by dirtyvagabond** [comment]
thanks @mattgillooly . chatted with @amalloy and was reminded that there's a proposed patch for this, but it's waiting on changes to the rhizome lib.

https://groups.google.com/forum/#!msg/drake-workflow/V0PypFEQqn8/Jyq8NeYrCzMJ
https://github.com/Factual/drake/commit/6ec6913886107a263ce11240a6e40f66bf89747e

**Comment by ztellman** [comment]
How on earth do I keep forgetting about this.  Okay, done by end of today.

**Comment by ztellman** [comment]
Okay @amalloy, changes to Rhizome are in place as of `0.2.3`.  Let me know if you have any questions.

**Comment by amalloy** [comment]
Looks good, except for one thing: the frame doesn't dispose itself when
closed (understandable, since your use case invilves reusing the same
frame), which means the JVM hangs around indefinitely. In order to actually
use this and cause the JVM to exit as normal, I have to do something like:
(let [frame (...)] (view-image frame ...) @done-promise (.dispose (first
@frame))), which obviously is relying on rhizome internals that I shouldn't
be touching. What would be best, from my perspective, is a way to set
DISPOSE_ON_CLOSE on the created frame.

**Comment by ztellman** [comment]
Okay, I was noticing that even when I called `.dispose()` I could still use the frame, which may or may not be some sort of implementation detail I can't rely on.  Can you confirm that if you *do* call `(-> f first .dispose)` it does what you expect?  If so, I'll add a `:dispose-on-close?` flag, which is what I had before but removed because it didn't seem to do anything.

**Comment by amalloy** [comment]
Yes, when I do that it does exactly what I want.

**Comment by ztellman** [comment]
Okay, pushed `0.2.4-SNAPSHOT`, if you confirm everything works as expected I'll make a new full release.

**Comment by amalloy** [comment]
A++, would use again.

**Comment by ztellman** [comment]
`0.2.4` is live.

**Comment by dirtyvagabond** [comment]
the develop branch has been updated with the ability to generate a PNG of the dependency graph (see help content below.)

@mattgillooly would you be willing to try this out and give us your stamp of approval?

```
   -g[arg]  Runs Drake in "graph" mode. Instead of
                    --graph[=arg]  executing steps, Drake just draws a graph of
                                   all the inputs and outputs in the workflow,
                                   with color-coding to indicate which will be
                                   run. The graph is saved to a file named
                                   drake.png in the current directory. Files
                                   which will be built are colored green, and
                                   those which were forced will be outlined in
                                   black as well. You can specify --graph=dot
                                   to output a .dot file instead of displaying
                                   the image and outputting a png.
```

**Comment by aboytsov** [comment]
Nice! :)

On Wed, Feb 25, 2015 at 4:21 PM, Aaron Crow <notifications@github.com>
wrote:

> the develop branch has been updated with the ability to generate a PNG of
> the dependency graph (see help content below.)
>
> @mattgillooly <https://github.com/mattgillooly> would you be willing to
> try this out and give us your stamp of approval?
>
>    -g[arg]  Runs Drake in "graph" mode. Instead of
>                     --graph[=arg]  executing steps, Drake just draws a graph of
>                                    all the inputs and outputs in the workflow,
>                                    with color-coding to indicate which will be
>                                    run. The graph is saved to a file named
>                                    drake.png in the current directory. Files
>                                    which will be built are colored green, and
>                                    those which were forced will be outlined in
>                                    black as well. You can specify --graph=dot
>                                    to output a .dot file instead of displaying
>                                    the image and outputting a png.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Factual/drake/issues/37#issuecomment-76094583>.
>



-- 
Artem.

**Comment by mattgillooly** [comment]
@dirtyvagabond - thanks!  This is better than my hacked-together script, and I will switch to using this immediately.

I was surprised by two things.  They don't necessarily feel like the *wrong* behavior, just things I didn't expect from the help text.

1.) On a project which has all its outputs up-to-date, no graph is produced.  I assumed, because of the color-coding described in the help text, that every project would produce a graph, and that in this case, the vertices would all be green.

2.) I was surprised that the PNG image was automatically opened in a viewer, rather than simply written to disk.  This is actually great for my current purposes, just surprising to me.  The viewer is more zoomed in than I would like, though, and I can't find a way to zoom it out, so I currently have to scroll to see the entire graph.  FWIW, the Preview app on my Mac scales the same PNG to fit on my screen.

**Comment by dirtyvagabond** [comment]
thanks @mattgillooly for demo'ing!

1) the idea here is to produce a graph of "what would Drake do if told to run right now with these targets".  so e.g. if you want "the entire graph as if no work has been done" you'd run with your targets as `+...` to indicate "force all". i think this is generally consistent from a usability perspective. i can augment the help message for this, and also call it out in the upcoming wiki docs to add clarity. does this seem reasonable to you?

2) i guess the last sentence in the help text hints at this for you (that by default we're gonna throw the image all up in yer grill). do you think that instead the default should be to only produce the file?

**Comment by amalloy** [comment]
I guess I never tried graphing when there was nothing to do: I expected it to produce a graph too! I'll see if I can fix that, because the current behavior is weird and surprising. As for sizing the image...man I have no idea how to do this stuff with Rhizome. Maybe it's better to just not attempt to display the image at all, since for most realistic workflows the image is probably going to be sized wrong anyway, and just let the user's image viewer handle that work.

**Comment by amalloy** [comment]
Okay, addressing (1) was easy; I've just made that change. Any suggestions on (2)? Is the current behavior (displaying an image badly) better or worse than not displaying the image at all? I don't think it's super-likely that we will get to "displaying an image well", so those are kinda the choices we have available.

**Comment by mattgillooly** [comment]
> 1) the idea here is to produce a graph of "what would Drake do if told to run right now with these targets".

@dirtyvagabond - This makes sense, and is consistent with the --print behavior, but I'm left wondering what the color coding accomplishes in that case.  If the graph only shows steps that would be run, then how would a non-green vertex ever appear?

>  Any suggestions on (2)? Is the current behavior (displaying an image badly) better or worse than not displaying the image at all?

@amalloy - I could be convinced either way.  In [pipely](https://github.com/swipely/pipely), we decided not to handle viewing, but I have had complaints from people who wish they had to do less typing to get their graphs to pop up.  PNG is a very stable interface, and nearly everyone has a viewer, so if you are drawing a cut-line, that would be a good one, IMO.

**Comment by dirtyvagabond** [comment]
@amalloy @mattgillooly sorry, i think i added unnecessary confusion for (1), by not understanding the actual requirements. to hopefully clarify: can we all agree that the produced diagram should always show the entire workflow, with colours used to indicate the state of that workflow? 

for (2) i guess i could be convinced either way as well. my vote, if forced, would be to produce only the file by default, and ask the user to provide the explicit flag to give Drake the permission to "take over my screen and throw an image all up in my grill"

**Comment by dirtyvagabond** [comment]
or for (1), @amalloy what's your full and clear spec?

**Comment by amalloy** [comment]
The full spec is what @mattgillooly expected: --graph will produce a graph of all steps in the workflow, regardless of which ones currently need to be run; it will use color coding to indicate which ones currently need running.

**Comment by amalloy** [comment]
And I like the idea of defaulting to just saving the image, with an option to display it on screen instead. I've already got the mechanics for `--graph=x`, so I can just add `--graph=all-up-in-my-grill`.

**Comment by dirtyvagabond** [comment]
sold!

**Comment by aboytsov** [comment]
Btw, are we doing this by running a process (vs becoming an X client)? I
think as long as it runs a separate process (and we probably need to
provide option to specify a shell command to do that), it doesn't violate
the spirit of Drake. I'd object to making Drake an X client though. I'd
also rather not do "all-up-in-my-grill" trick as it seems rather arbitrary.
My proposal would be:

1) --graph=filename (creates filename)
2) --graph-process="display --rotate 90"
3) --graph (no argument: creates a $TEMPFILE and runs "display --rotate 90
$TEMPFILE" via $SHELL)

There may be some issues with overloading --graph but I can't think of any.
Let me know what you think.

Also, with regards to what exactly is being graphed, I think there's value
in both showing the subset of steps scheduled for execution separately, and
in context of the overall workflow. The spec might mention something along
these lines, but I don't remember well. The most flexible way to do that
would be to allow specification of the graph scope as a target set
alongside the main target set, for example:

drake --graph A B C (shows what targets will be built to make A, B and C)
drake --graph --graph-scope="+..." A B C (shows the same targets but in the
context of the overall workflow, targets not shown in the previous example
could be grey or something?)
drake --graph --graph-scope="+A +B +C" A B C (shows the same targets but in
the context of forcibly rebuilding A, B and C allowing to see which parts
of the workflow are NOT being rebuilt)

Implementing it this way can make graphing a powerful workflow debugging
tool.

Yours,
  Artem.

On Thu, Feb 26, 2015 at 12:40 PM, Aaron Crow <notifications@github.com>
wrote:

> sold!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Factual/drake/issues/37#issuecomment-76266930>.
>



-- 
Artem.

**Comment by amalloy** [comment]
I don't want to spin up another process to display an image: I'm already a little sad that we have to shell out to `dot` to create the image. Defaulting to `display` might sound fine if you're on a mac, but other operating systems don't have that - why not default to `xdg-open`, or to `mspaint` instead? I just don't want to add another external dependency, when the JVM already has portable facilities for displaying images in-process. I see "output PNG file" as the option for when you want to do something flexible or complicated; you can output to a file, then do whatever fancy stuff you want with an image viewer. Displaying the image directly to your screen is a shortcut for when you want to do the simplest thing, so it's fine if it doesn't include a bunch of options.

As for the graph scope stuff...sure, that sounds cool, but I couldn't figure out how to do something like that using the information that drake gets from parsing the workflow and deciding what to run: how do I tell the difference between a step that's not being built because it was "outside of scope", vs one that was in scope but didn't need to be built? The generated graph already distinguishes those steps that are force-built vs needed to be rebuilt, so I think the only thing your proposal is asking for is a way to make a specific set of the white nodes disappear from the graph entirely; that's not super-useful IMO, because the user can choose to just ignore the nodes they don't want to look at. But if you want to add a more delicate node-selection feature, patches are welcome, of course!

**Comment by dirtyvagabond** [comment]
This was finalized and recently released with Drake 0.1.7
